### PR TITLE
adding gen 2 instructions for mySQL / postgreSQL support

### PR DESCRIPTION
**Pull request by renebrandel**
### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+// highlight-start
+rdsSchema.addQueries({
+  searchTodoBasedOnGeolocation: a.query()
+    .arguments({
+      lat: a.float(),
+      long: a.float(),
+      radiusInMeters: a.float()
+    })
+    .returns(a.ref("Todo").array())
+    .handler(sql`
```
Comment (stocaaro) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:384]: I'm unsure what the `sql` template string function is doing for us here. Is this a function that our library provides that should be added to the import above? Is this coming from somewhere else? What behavior does it provide?

Is this most similar to one of the following?
- https://github.com/andywer/squid?tab=readme-ov-file
- https://www.npmjs.com/package/sql-template-strings
- https://www.npmjs.com/package/sql-template-tag

I thought I might find this idiom in Prisma, but they appear to support raw sql as plain strings.

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+// highlight-start
+rdsSchema.addQueries({
+  searchTodoBasedOnGeolocation: a.query()
+    .arguments({
+      lat: a.float(),
+      long: a.float(),
+      radiusInMeters: a.float()
+    })
+    .returns(a.ref("Todo").array())
+    .handler(a.handler.inlineSql`
```
Comment (stocaaro) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:222]: This also reads like a template string array function. Probably the long form of what is implied above. I would suppose that this is meant to indicate that we need a pattern to protect against sql injection. Is there a library that we intend to consume for this protection? If not consuming, are we patterning this behavior off of some other specific library?

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+<Block name="Amplify console"></Block>
+
+</BlockSwitcher>
+
+## Step 2 - Generate database schema
+
+Run the following command to generate the Data schema with your database connection information.
+
+```bash
+npx amplify generate schema-from-database --connection-string MYSQL_CONNECTION_STRING --out amplify/data/schema.rds.ts
```
Comment (sundersc) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:49]: Changed the option name after the most recent API review.

```suggestion
npx amplify generate schema-from-database --connection-uri-secret MYSQL_CONNECTION_STRING --out amplify/data/schema.rds.ts
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        location,
+        ST_MakePoint(:lat, :long)
+      ) <= :radiusInMeters`
+    )
+    .authorization([a.allow.public()])
+})
+// highlight-end
+
+// highlight-start
+rdsSchema.setSqlStatementFolderPath(path.join(__dirname, 'aurora-statements'))
```
Comment (renebrandel) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:218]:
```suggestion
rdsSchema.setSqlStatementFolderPath('./aurora-statements/'))
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+## Step 1 - Set secrets for database connection
+
+First, provide all the database connection information as secrets, you can use the Amplify sandbox's secret functionality to set them or go to the Amplify console to set secrets in a shared environment:
+
+<BlockSwitcher>
+<Block name="Sandbox">
+
+```bash
+npx amplify sandbox secret add MYSQL_CONNECTION_STRING
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:34]:
```suggestion
npx amplify sandbox secret set MYSQL_CONNECTION_STRING
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+const ddbSchema = a.schema({
+   Note: a.model({
+     content: a.string(),
+     todoId: a.id()
+     todo: a.belongsTo("Todo").references("todoId")
+   })
+})
+
+const combinedSchema = a.schema.combine([rdsSchema, ddbSchema])
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:178]: New DX:

```suggestion
const modifiedRds = rdsSchema.relationships((models) => [
  models.Todo.addRelationships({
    // will be: a.hasMany("Comment", "todo_id")
    comments: a.hasMany("Comment").references("todo_id"),
    // will be: a.hasMany("Note", "todoId")    
    notes: a.hasMany("Note").references("todoId")
  })
]);
// highlight-end

const ddbSchema = a.schema({
   Note: a.model({
     content: a.string(),
     todoId: a.id()
     todo: a.belongsTo("Todo").references("todoId")
   })
})

const combinedSchema = a.schema.combine([modifiedRds, ddbSchema])
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+// highlight-start
+rdsSchema.addQueries({
+  searchTodoBasedOnGeolocation: a.query()
+    .arguments({
+      lat: a.float(),
+      long: a.float(),
+      radiusInMeters: a.float()
+    })
+    .returns(a.ref("Todo").array())
+    .handler(a.handler.inlineSql`
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:205]:
```suggestion
    .handler(a.handler.inlineSql(`
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+        location,
+        ST_MakePoint(:lat, :long)
+      ) <= :radiusInMeters`
+    )
+    .authorization([a.allow.public()])
+})
+// highlight-end
+
+// highlight-start
+rdsSchema.setSqlStatementFolderPath('./aurora-statements/'))
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:218]: No longer supported. `a.handler.sqlReference('./relativePathToFile.sql')` is how we specify a path for the file containing SQL statements
```suggestion
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+  updateTodoBasedOnLocation: a.mutation()
+    .arguments({
+      lat: a.float(),
+      long: a.float(),
+      radiusInMeters: a.float(),
+      todoChange: a.ref("Todo")
+    })
+    .returns(a.ref("Todo"))
+   .handler(a.handler.sqlReference())
+    .authorization([a.allow.public()])
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:230]:
```suggestion
const modifiedRds = rdsSchema.addMutation({
  updateTodoBasedOnLocation: a.mutation()
    .arguments({
      lat: a.float(),
      long: a.float(),
      radiusInMeters: a.float(),
      todoChange: a.ref("Todo")
    })
    .returns(a.ref("Todo"))
    .handler(a.handler.sqlReference('./updateTodoBasedOnLocation.sql'))
    .authorization([a.allow.public()])
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+const schema = a.schema({
+  Todo: a.model({
+    content: a.string()
+    isDone: a.boolean()
+  })
+})
+
+const combinedSchema = a.schema.combine([
+  schema,
+  rdsSchema
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:243]:
```suggestion
  modifiedRds
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+
+### Apply authorization rules
+
+```ts title="amplify/data/resource.ts"
+import { a, defineData } from '@aws-amplify/backend'
+import { schema as rdsSchema } from './schema.rds.ts'
+
+// Add an authorization rule to the schema
+// highlight-next-line
+rdsSchema.models.Todo.authorization([a.allow.public()])
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:300]:
```suggestion
const modifiedRds = rdsSchema.setAuthorization((models) => [
  models.todo.authorization([a.allow.public()]),
  models.todo.fields.content.authorization([a.allow.private()]),
]);
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+const schema = a.schema({
+  Todo: a.model({
+    content: a.string()
+    isDone: a.boolean()
+  })
+})
+
+const combinedSchema = a.schema.combine([
+  schema,
+  rdsSchema
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:311]:
```suggestion
  modifiedRds
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+### Rename inferred model and field names
+
+```ts title="amplify/data/resource.ts"
+import { a, defineData } from '@aws-amplify/backend'
+import { schema as rdsSchema } from './schema.rds.ts'
+
+// Rename models or fields to be more idiomatic for frontend code
+// highlight-start
+rdsSchema.models.todos.renameTo("Todo")
+rdsSchema.models.todos.fields.title.renameTo("content")
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:329]: Same DX changes as above

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+// highlight-start
+rdsSchema.addQueries({
+  searchTodoBasedOnGeolocation: a.query()
+    .arguments({
+      lat: a.float(),
+      long: a.float(),
+      radiusInMeters: a.float()
+    })
+    .returns(a.ref("Todo").array())
+    .handler(sql`
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:366]:
```suggestion
    .handler(a.handler.inlineSql(`
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+      SELECT *
+      FROM todos
+      WHERE ST_DistanceSphere(
+        location,
+        ST_MakePoint(:lat, :long)
+      ) <= :radiusInMeters
+    `)
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:373]:
```suggestion
    `))
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+
+### Invoke SQL statements with custom queries and mutations
+
+```ts title="amplify/data/resource.ts"
+import { a, defineData } from '@aws-amplify/backend'
+import { schema as rdsSchema } from './schema.rds.ts'
+
+// Add custom mutations or queries that execute SQL statements
+// highlight-start
+rdsSchema.addQueries({
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:358]:
```suggestion
const modifiedRds = rdsSchema.addQueries({
```

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+const schema = a.schema({
+  Todo: a.model({
+    content: a.string()
+    isDone: a.boolean()
+  })
+})
+
+const combinedSchema = a.schema.combine([
+  schema,
+  rdsSchema
```
Comment (iartemiev) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:387]:
```suggestion
  modifiedRds
```

**Review by iartemiev — CHANGES_REQUESTED** [review]

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+If your database is behind a VPC, we'll autodetect the VPC configuration based on your connection string. This happens via a hostname lookup within your account. The sandbox profile in your account needs to have XYZ permissions to do a VPC lookup and connect to the VPC. (If your database is in a private subnet, we'll provision a Lambda to retrieve the database schema in order for us to generate the TypeScript data model.) TODO: Chris / Ed to connect on how best to implement this.
+
+</Callout>
+
+This creates a new **schema.rds.ts** with a schema reflecting the types of your database. Import the schema to your **amplify/data/resource.ts** file and
+
+```ts
+import { a, defineData } from '@aws-amplify/backend';
+// highlight-next-line
+import { schema as auroraSchema } from './schema.rds.ts';
```
Comment (phani-srikar) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:63]: Is there any particular reason for using `auroraSchema` versus a more general `sqlSchema`?

File: src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+
+Customers need to update the `amplify.yml` to generate the DB schema if the DB is in a different VPC. Etc Etc. TK
+
+TK: npx amplify generate config --connection-string $BRANCH_CONNECTION_STRING
+
+## Add relationships between tables
+
+```ts
+import { a, defineData } from '@aws-amplify/backend'
+import { schema as rdsSchema } from './schema.rds.ts'
```
Comment (phani-srikar) [inline src/pages/gen2/build-a-backend/data/connect-to-existing-database/index.mdx:167]: here and in other snippets, specifying `.ts` gives the following error in my VS Code IDE with default settings:
Screenshot 2024-04-17 at 12 22 13 PM
We could skip that extension and that resolves it.

**Pull request closed without merging**